Restructure CMake code to allow external usage

nicolasfella opened the pull request:
- Add a CMake config file to allow finding the library via find_package
- Extract the registration handling into gr_generate_block_instantiations function
- Move the function into a separate CMake file
- Create an interface library for gr_registry.hpp to allow exporting and automatic include dir propagation
- Install generator program

This allows to use the library and the generation function from both find_package and FetchContent

See https://github.com/fair-acc/gnuradio4/issues/517

nicolasfella committed:
Restructure CMake code to allow external usage

- Add a CMake config file to allow finding the library via find_package
- Extract the registration handling into gr_generate_block_instantiations function
- Move the function into a separate CMake file
- Create an interface library for gr_registry.hpp to allow exporting and automatic
include dir propagation
- Install generator program

This allows to use the library and the generation function from both
find_package and FetchContent

nicolasfella commented:
Example project that uses this to define a block: https://github.com/nicolasfella/gr4-block-user/

RalphSteinhagen reviewed (approved):
@nicolasfella thanks a lot for the clean-up and restructuring! Looks very nice.

Would you know whether it's possible to automatically regenerate (i.e. 'delete' and 're-create')  the `.cpp` files in response to changes in the defining `.hpp` file?

The practical use case is that a user 
 * adds or removes a template specialisation (e.g. `[ int ]` -> `[ int double ]` -> `[ float ]`, or
 * changes the template name (e.g. `Block3` -> `Block3a`).
 
 A subsequent build detects the changes but leaves the old (often incorrect) `.cpp` in tact. also just rerunning `cmake` doesn't seem to solve. Could this be solved for regular build calls? :thinking: 
 
 Maybe this question is also for @ivan-cukic

nicolasfella committed:
Also match dummy file when globbing for old cpp files

nicolasfella committed:
Automatically re-run CMake when a header changes

The code generator runs at configure time, so any time one of the headers changes we want to re-run the generation

nicolasfella commented:
I added a change that automatically re-runs cmake when one of the headers changes. With that (and a small tweak to the cleanup glob) it seems to work as expected

RalphSteinhagen commented:
@nicolasfella great! 

Would it be possible to limit the rebuild only to the header files and dependent `.cpp` files that have actually changed? The implementation you posted worked but would imply that if I have 20-30 header files and only one of them has changed, all blocks in the sub-module would need to be rebuilt. 

Hope this isn't asking for too much. :see_no_evil:

nicolasfella commented:
I don't think it's _easily_ possible to avoid this. The problem is that when any header changes we re-run CMake, and that causes the generation for _all_ headers to run and the source file is updated and needs recompilation.

To avoid that we probably could manually check the file modification timestamp and compare that to a saved value to only re-run the generation.

nicolasfella committed:
Only run generator if header file changed

Otherwise a change in one header causes all of them to be reprocessed

nicolasfella commented:
I added a change that does such a timestamp check. It seems to work, but it's definitely not perfect as it won't detect e.g. if the source file was removed and therefore needs to be regenerated

RalphSteinhagen commented:
@nicolasfella thanks for the update. I think the last commit broke the build and the generator isn't been called at all. Could you check again? Thanks in advance.

> [..] not perfect as it won't detect e.g. if the source file was removed and therefore needs to be regenerated

I think that is OK. Most users won't tinker in the build directory and delete some random files. The global 'clean' and 'rebuild' step should work though. The `ccache` should catch most of the compile-time overhead.

nicolasfella commented:
ah, it was broken for the initial cmake run because there was no cached timestamp yet.

Please try again

RalphSteinhagen commented:
@nicolasfella much better. :+1: 

Hope this is not too nit-picky: Could the following cmake warning be fixed, for example:
  ```cmake
  modified_include/gnuradio-4.x/basic/block3.hpp
  CMake Warning (dev) at build/BlockLibMacros.cmake:28 (set):
    implicitly converting 'String' to 'STRING' type.
  Call Stack (most recent call first):
    CMakeLists.txt:69 (gr_generate_block_instantiations)
    blocks/basic/CMakeLists.txt:16 (create_block_library)
  This warning is for project developers.  Use -Wno-dev to suppress it.
  ```
  
  Not major. If you read this within the next ~half an hour would be great if this could be fixed. Otherwise I'd merge it as is.
  
  Thanks again for your expert help. It's clear to see the difference of a C++/cmake master as you are and some debutant like me. :smile:

nicolasfella commented:
Fixed :)

RalphSteinhagen commented:
Splendid!
